### PR TITLE
chore: update TypeScript to v5

### PR DESCRIPTION
**Pull request by fisker**

### CODE REVIEW

File: src/common.ts [hunk 1]
```diff
@@ -831,3 +831,9 @@ export function classifyIdentifier(parser: ParserState, context: Context, t: Tok
 
   if (!isValidIdentifier(context, t)) report(parser, Errors.Unexpected);
 }
+
+/**
+ * A helper function to check parser state, workaround TS2678 error
+ * @see https://github.com/microsoft/TypeScript/issues/9998
+ */
+export const is = (value: number | boolean) => value;
```
Comment (fisker) [inline src/common.ts:839]: It's annoying that TS narrowing down the type of `parser.token` incorrectly, [issue](https://github.com/microsoft/TypeScript/issues/9998), this function prevents that.

Do you have a better idea?

File: src/parser.ts [hunk 2]
```diff
@@ -5528,7 +5533,7 @@ function parseSpreadOrRestElement(
     );
 
     const { token, tokenPos, linePos, colPos } = parser;
-
+    // @ts-expect-error BUG?
     if (token === Token.Assign && token !== closingToken && token !== Token.Comma) {
```
Comment (fisker) [inline src/parser.ts:5537]: Is this a mistake?

File: src/common.ts (hunk 1 above)
Comment (3cp) [reply src/common.ts:839]: I will have a play.

File: src/parser.ts (hunk 2 above)
Comment (3cp) [reply src/parser.ts:5537]: line 5485 `if (token !== Token.Assign && token !== closingToken && token !== Token.Comma) {`
That's more like it.

I tested this line with `if (token === Token.Assign) {`, all tests still pass. Please fix the code. Thx!

File: src/common.ts (hunk 1 above)
Comment (3cp) [reply src/common.ts:839]: Is this a tsc v5 bug? tsc v4 has no such issue, but tsc v4 fails all the places.

Comment (3cp) [reply src/common.ts:839]: I would suggest to hold off tsc v5 before we have a proper solution.

File: src/parser.ts (hunk 2 above)
Comment (fisker) [reply src/parser.ts:5537]: But I use `if (token === Token.Assign)`, test also pass.

Comment (3cp) [reply src/parser.ts:5537]: #271

**Comment by 3cp** [comment]
I have seen weird ts v5.5 typing issue in other projects.
ts v5 seems trying to be too smart.

**Comment by fisker** [comment]
Maybe this solution is acceptable? https://github.com/microsoft/TypeScript/issues/9998#issuecomment-235963457

**Comment by 3cp** [comment]
I think it's a ts bug, it assumes our `nextToken(parser, ...)` didn't mutate `parser.token`.
But `parser.token =` is right there in the implementation of `nextToken`.

**Comment by 3cp** [comment]
I can reproduce the issue https://www.typescriptlang.org/play/?ts=5.5.3#code/JYOwLgpgTgZghgYwgAgCoQM5mQbwFDKHIwA2cA5gFzIAMyAPsgIwDceAvnnjAK4gJhgAexDIAtjzBxIACkhZq6LAEpcBIvLAA6UhWQBeZm054EIrMjiLM2QzmJkqtZOzZ5gMZDLg7HB-YY0qvhE4pLSEN7KbKEeXj665P6GTMHqoYRmIBhCJBBaJELkMgDkiSUANJa+FNHpLhx4QA

But the same typing error also pops up with ts v4.

**Comment by 3cp** [comment]
Had a read of the ts github issue. It sounds like ts would not address this issue anytime soon.

If you have time to apply the `parser.currentToken() or getToken()` solution, please go ahead :-)

**Comment by 3cp** [comment]
I am also very confused why we didn't have this issue with ts v4 and v3 before.

**Comment by fisker** [comment]
Done.

**Review by 3cp — APPROVED** [review]

**Pull request merged**